docs/architecture/labeling.md

```diff
@@ -0,0 +1,250 @@
+# Labeling
+
+## Requirements
+
+- User can label every top-level entity like Application, Runtime
+- User can search Application/Runtime by the specific label and its value
+- User can find out about all label keys used in the given tenant
+- User can define validation rules for a label with a given key but it is optional
+- There is one special label: **Scenarios**, that has additional requirements:
+    - Every Application is labeled with **Scenarios**
+    - By default, **Scenarios** has one possible value: **default**
+    
+    
+## Proposed solution
+
+1. Application or Runtime can be tagged with a label that consists of **key** and **value**.
+1. A label can have related LabelDefinition when a user can define JSON schema for validation of values.
+Thanks to that our API is extremely simple and we can treat every label in the same way.
+Even UI developers can benefit from it because there are libraries that render proper JS component
+on the JSON schema basis. See [this library](https://github.com/json-editor/json-editor).
+
+2. For the given label you can provide only one value. Still, a value can be JSON array, depending on LabelDefintions schema.
+3. LabelDefinition is optional, but it will be created automatically if the user adds a label for which LabelDefinition does not exist.
+ 
+### API Changes
+
+1. Extend GraphQL API with  mutations and queries for a new type: **LabelDefinition** 
+
+```graphql
+type LabelDefinition {
+    key: String!
+    schema: String
+}
+
+input LabelDefinitionInput {
+    key: String!
+    schema: String
+}
+
+
+type Query {
+    labelDefinitions: [LabelDefinition!]!
+    labelDefinition(key: String!): LabelDefinition
+}
+
+type Mutation {
+    createLabelDefinition(in: LabelDefinitionInput!): LabelDefinition!
+    updateLabelDefinition(in: LabelDefinitionInput!): LabelDefinition!
+    deleteLabelDefinition(key: String!, deleteRelatedLabels: Boolean=false): LabelDefinition!
+}
+```
+
+The LabelsDefinition key has to be unique for a given tenant. A schema defines JSON schema used when a user adds a label
+to the Application or Runtime. 
+
+1. In JSON schema user can define if given label accepts one or many values. Because of that, we have to change
+or API and allow to specify only one value for a given label. This value can contain many elements, depending on LabelDefinition's schema.  
+Change from
+```graphql
+addApplicationLabel(applicationID: ID!, key: String!, values: [String!]!): Label!
+```
+to:
+```graphql
+setApplicationLabel(applicationID: ID!, key: String!, value: String!): Label!
```

Comment:
Shouldn't we use `Any` for `value`?

Reply:
For whole label probably `Any` scalar would be the best guess, because is converted to `interface{}`. If we would like to manipulate these values in future, it will be easier to not unmarshal object from string.

https://gqlgen.com/reference/scalars/#any

Reply:
for setting label value and LabelDefinition schema we use JSON which is equivalent to Any